src/devices_models/devices/common/add_variable.jl

```diff
@@ -1,3 +1,150 @@
+    end
+    if !isnothing(variable_name) && !isnothing(variable_names)
+        throw(ArgumentError("variable_name and variable_names cannot both be set"))
+    end
+    if !isnothing(variable_name)
+        variable_names = [variable_name]
+    end
+
+    return AddVariableInputs(
+        variable_names,
+        binary,
+        expression_name,
+        sign,
+        devices_filter_func,
+        initial_value_func,
+        lb_value_func,
+        ub_value_func,
+    )
+end
+
+function make_active_power_add_variable_inputs(
+    ::Type{T},
+    ::PSIContainer,
+) where {T <: PSY.Device}
+    error("make_active_power_add_variable_inputs is not implemented for $T")
+end
+
+function make_reactive_power_add_variable_inputs(
+    ::Type{T},
+    ::PSIContainer,
+) where {T <: PSY.Device}
+    error("make_reactive_power_add_variable_inputs is not implemented for $T")
+end
+
+function make_commitment_add_variable_inputs(
+    ::Type{T},
+    ::PSIContainer,
+) where {T <: PSY.Device}
+    error("make_commitment_add_variable_inputs is not implemented for $T")
+end
+
+function activepower_variables!(
+    psi_container::PSIContainer,
+    devices::IS.FlattenIteratorWrapper{T},
+) where {T <: PSY.Component}
+    inputs = make_active_power_add_variable_inputs(T, psi_container)
+    add_variables!(psi_container, devices, inputs)
+end
+
+function reactivepower_variables!(
+    psi_container::PSIContainer,
+    devices::IS.FlattenIteratorWrapper{T},
+) where {T <: PSY.Component}
+    inputs = make_reactive_power_add_variable_inputs(T, psi_container)
+    add_variables!(psi_container, devices, inputs)
+end
+
+function commitment_variables!(
+    psi_container::PSIContainer,
+    devices::IS.FlattenIteratorWrapper{T},
+) where {T <: PSY.Component}
+    inputs = make_commitment_add_variable_inputs(T, psi_container)
+    add_variables!(psi_container, devices, inputs)
+end
+
+"""
+Add variables to the PSIContainer from type-specific inputs.
+"""
+function add_variables!(
+    psi_container::PSIContainer,
+    devices::IS.FlattenIteratorWrapper{T},
+    inputs::AddVariableInputs,
+) where {T <: PSY.Component}
+    _add_variables!(psi_container, devices, inputs)
+end
+
+function add_variables!(
+    psi_container::PSIContainer,
+    devices::IS.FlattenIteratorWrapper{T},
+    inputs::Vector{AddVariableInputs},
+) where {T <: PSY.Component}
+    for input in inputs
+        _add_variables!(psi_container, devices, input)
+    end
+end
+
+function _add_variables!(
+    psi_container::PSIContainer,
+    devices::IS.FlattenIteratorWrapper{T},
+    inputs::AddVariableInputs,
+) where {T <: PSY.Component}
+    variable_names = inputs.variable_names
+    binary = inputs.binary
+    expression_name = inputs.expression_name
+    sign = inputs.sign
+    initial_value_func = inputs.initial_value_func
+    lb_value_func = inputs.lb_value_func
+    ub_value_func = inputs.ub_value_func
+
+    if !isnothing(inputs.devices_filter_func)
```

Comment:
Am I interpreting comments correctly such that I should do the following?

```
filter_func = nothing
if isnothing(inputs.devices_filter_func)
    if T <: PSY.Device
        filter_func = x -> PSY.get_available(x)
    end
else
    filter_func = inputs.filter_func
end
```

Reply:
yes, that's reasonable